### PR TITLE
Add image building profile

### PR DESCRIPTION
**Pull request by tych0**
With this I can build an image using buildkit.

It's requesting the host networking folks because i just ran `make protoc` and it updated their protos as well. I can drop those hunks from the patches if it's annoying.

### CODE REVIEW

**Comment by hechaoli** [comment]
I'm file with the change in `vpc/api/vpc.pb.go`. Nothing but a a version bump. Will leave the rest of the review to Kyle.

File: executor/runtime/docker/seccomp/titus.libsonnet [hunk 1]
```diff
@@ -64,4 +64,21 @@
...
+    names: [
+      'clone',
+      'fsconfig',
+      'fsmount',
+      'fsopen',
+      'fspick',
+      'keyctl',
+      'mount',
+      'move_mount',
+      'pivot_root',
```
Comment (solarkennedy) [inline executor/runtime/docker/seccomp/titus.libsonnet:78]: You are right, keyctl/pivot were added to fuse *only* so we could hack it and piggyback on that profile for image building.

You are hitting on the very reason for a new profile! :)
Lets remove those other syscalls from fuse in another pr

File: Makefile [hunk 2]
```diff
@@ -138,20 +138,16 @@ push-titus-agent: titus-agent
 
 ## Protobuf and source code generation
 
-PROTO_DIR     = vendor/github.com/Netflix/titus-api-definitions/src/main/proto
+PROTO_DIR     = api
```
Comment (solarkennedy) [inline Makefile:141]: You mention a script to get the closed source protos in the commit description, but I can't find it? Did you git add?

File: executor/runtime/docker/capabilities_linux.go [hunk 3]
```diff
@@ -103,10 +134,11 @@ func setupAdditionalCapabilities(c runtimeTypes.Container, hostCfg *container.Ho
 		hostCfg.Sysctls["net.ipv4.conf.all.arp_ignore"] = "1"
 	}
 
+	logrus.Info("using seccomp profile ", seccompProfile)
```
Comment (solarkennedy) [inline executor/runtime/docker/capabilities_linux.go:137]: debug is fine here

File: executor/runtime/docker/capabilities_linux.go [hunk 4]
```diff
@@ -65,9 +83,22 @@ func setupAdditionalCapabilities(c runtimeTypes.Container, hostCfg *container.Ho
 		seccompProfile = "fuse-container.json"
 	}
 
+	// the image building apparmor+seccomp profiles are a strict superset
+	// of the fuse ones, but fuse allows access to /dev/fuse, so these need
+	// to come after fuse in the event that people asked for both
```
Comment (solarkennedy) [inline executor/runtime/docker/capabilities_linux.go:88]: Its only a strict subset *now*.

This is a new API. Let's start strict.
Please error out when someone asks for both.

File: executor/runtime/docker/capabilities_linux.go [hunk 5]
```diff
@@ -65,9 +83,22 @@ func setupAdditionalCapabilities(c runtimeTypes.Container, hostCfg *container.Ho
...
+	if hasCapability(containerCapabilities, titusAPI.ContainerCapability_ContainerCapabilityImageBuilding) {
+		if _, ok := addedCapabilities[SYS_ADMIN]; !ok {
+			hostCfg.CapAdd = append(hostCfg.CapAdd, SYS_ADMIN)
+		}
+
+		seccompProfile = "image-building.json"
+		apparmorProfile = "docker_image_building"
+	}
+
 	if c.SeccompAgentEnabledForPerfSyscalls() {
```
Comment (solarkennedy) [inline executor/runtime/docker/capabilities_linux.go:99]: We'll fix this soon.
Now that we have *much* better way of wrangling our seccomp profiles, this won't be so hard anymore to enable in all of them by default. (in another PR)

File: executor/runtime/docker/capabilities_test.go [hunk 6]
```diff
@@ -58,14 +65,80 @@ func TestOverrideFuseProfile(t *testing.T) {
...
+
+	assert.NoError(t, setupAdditionalCapabilities(c, &hostConfig, runtimeTypes.MainContainerName))
+
+	assert.Contains(t, hostConfig.CapAdd, "SYS_ADMIN")
+	assert.Len(t, hostConfig.CapDrop, 0)
+	assert.Len(t, hostConfig.SecurityOpt, 2)
+
+	// image building is a superset of fuse, so we use that profile when
+	// both are supplied
+	assert.Contains(t, hostConfig.SecurityOpt, "apparmor:docker_image_building")
```
Comment (solarkennedy) [inline executor/runtime/docker/capabilities_test.go:135]: You are going above and beyond here.
For now lets kiss and only handle one capability at a time here.

File: executor/runtime/types/pod.go [hunk 7]
```diff
@@ -224,6 +225,42 @@ func (c *PodContainer) Capabilities() *corev1.Capabilities {
...
+
+		titusCap := titusAPI.ContainerCapability(c)
+		capabilities = append(capabilities, titusCap)
+
+		if titusCap == titusAPI.ContainerCapability_ContainerCapabilityFUSE {
+			hasFuse = true
+		}
+	}
+
+	// all containers in pods with the fuse-enabled annotation have the fuse capability
```
Comment (solarkennedy) [inline executor/runtime/types/pod.go:256]: This is a nice way to do it!

**Comment by solarkennedy** [comment]
I hope you didn't spend too much time fighting overlay.
I had already known what does and does not work from here: https://stash.corp.netflix.com/projects/TN/repos/titus-integration-tests/pull-requests/514/overview
But you came to the same conclusions, which is great!

**Review by solarkennedy — APPROVED** [review]
This is great.
Thanks for splitting up the commits.
Also thank you for being the first one to deal with the new proto stuff.

File: executor/runtime/types/pod.go [hunk 8]
```diff
@@ -224,6 +225,42 @@ func (c *PodContainer) Capabilities() *corev1.Capabilities {
...
+	if ok && val != "" {
+		caps = strings.Split(val, ",")
+	}
+
+	hasFuse := false
+	for _, cStr := range caps {
+		c, ok := titusAPI.ContainerCapability_value["ContainerCapability"+cStr]
+		if !ok {
+			logrus.Errorf("unknown container capability %s", cStr)
+			return nil, fmt.Errorf("unknown container capability %s", cStr)
```
Comment (solarkennedy) [inline executor/runtime/types/pod.go:245]: Why not return a real error here?
We can be strict.

File: Makefile (hunk 2 above)
Comment (tych0) [reply Makefile:141]: It's inline with the protos, I guess it makes more sense to live in `scripts/`? I can move it there.

File: executor/runtime/types/pod.go (hunk 8 above)
Comment (tych0) [reply executor/runtime/types/pod.go:245]: Not sure what you mean here, can you elaborate?

**Comment by tych0** [comment]
> I had already known what does and does not work from here: https://stash.corp.netflix.com/projects/TN/repos/titus-integration-tests/pull-requests/514/overview

Yeah, I knew it as well in principle. In practice it took a while to remind myself because the error from buildkit was so bad. But I hadn't seen your PR there at all, I made a very similar one, just using a tmpfs: 

https://stash.corp.netflix.com/projects/TN/repos/titus-integration-tests/commits/173b02319e67bf6a9863f39cf83d73ccf4a4acf5
https://stash.corp.netflix.com/projects/TN/repos/docker-files/commits/baaf48b703b4daec051d0a9b5d8300687f59efec

I did use figuring this out as an excuse to write: https://github.com/tych0/kernel-utils/blob/master/why-did-mount-fail.bt which is a tool I've always wanted.

File: executor/runtime/types/pod.go (hunk 8 above)
Comment (solarkennedy) [reply executor/runtime/types/pod.go:245]: Oh sorry I misread

**Pull request merged**